pyquil/reference_simulator.py

```diff
@@ -246,8 +265,19 @@ def expectation(self, operator: Union[PauliTerm, PauliSum]):
         raise NotImplementedError("To implement")
 
     def reset(self) -> 'AbstractQuantumSimulator':
-        self.density.fill(0)
-        self.density[0, 0] = complex(1.0, 0)
+        """
+        Resets the initial state of ReferenceDensitySimulator.
+
+        If the flag ``self.density_is_set`` is True then the self.density is reset to the user
+        specified self.initial_density. Otherwise it is set to the all zeros state.
+
+        :return: ``self`` to support method chaining.
+        """
+        if self.density_is_set:
```

Comment:
See above. This could just be `if self.initial_state: ...`. Or if you go the other route and extend this function to take `initial_state`, then it would just be `if initial_state: ...`.

Reply:
Good idea! I took this idea and used it to simplify the `set_density` and `reset` code.

- [x] todo